bundle/config/root.go

```diff
@@ -71,54 +79,149 @@ func Load(path string) (*Root, error) {
 	}
 
 	var r Root
-	err = yaml.Unmarshal(raw, &r)
+	v, err := yamlloader.LoadYAML(path, bytes.NewBuffer(raw))
 	if err != nil {
 		return nil, fmt.Errorf("failed to load %s: %w", path, err)
 	}
 
-	if r.Environments != nil && r.Targets != nil {
-		return nil, fmt.Errorf("both 'environments' and 'targets' are specified, only 'targets' should be used: %s", path)
-	}
+	// Normalize dynamic configuration tree according to configuration type.
+	v, diags := convert.Normalize(r, v)
 
-	if r.Environments != nil {
-		//TODO: add a command line notice that this is a deprecated option.
-		r.Targets = r.Environments
+	// Convert normalized configuration tree to typed configuration.
+	err = convert.ToTyped(&r, v)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load %s: %w", path, err)
 	}
 
+	r.diags = diags
+
+	// Store dynamic configuration for later reference (e.g. location information on all nodes).
+	r.value = v
+
 	r.Path = filepath.Dir(path)
-	r.SetConfigFilePath(path)
+	// r.SetConfigFilePath(path)
 
-	_, err = r.Resources.VerifyUniqueResourceIdentifiers()
+	// _, err = r.Resources.VerifyUniqueResourceIdentifiers()
 	return &r, err
 }
 
+func (r *Root) initializeValue() {
```

Comment:
shall we name it `fromTyped` or similar. Initialize sounds a bit misleading when reading the code